diesel/src/sqlite/connection/statement_iterator.rs

```diff
@@ -12,16 +12,75 @@ pub struct StatementIterator<'stmt, 'query> {
     field_count: usize,
 }
 
+impl<'stmt, 'query> StatementIterator<'stmt, 'query> {
+    #[cold]
+    fn handle_duplicated_row_case(
+        outer_last_row: &mut Rc<RefCell<PrivateSqliteRow<'stmt, 'query>>>,
+        column_names: &mut Option<Rc<[Option<String>]>>,
+        field_count: usize,
+    ) -> Option<QueryResult<SqliteRow<'stmt, 'query>>> {
```

Comment:
Curious what this kind of duplication handling is needed for. Is this a SQLite or SQLite driver bug?

Reply:
The reasoning here is a bit "complicated". The main issue here is that `SQLite` interface returns something like an iterator, but not exactly an iterator according to rusts definition. The main difference here is that `Iterator::next` is defined as `fn next(&mut self) -> Option<Self::Item>`. That implies that each item is independent from the underlying iterator, so it cannot hold a reference to the original iterator at all.

Now sqlites interface for prepared statements requires multiple functions to setup a statement. It boils down to the following steps:
1. Create a prepared statement itself by using [`sqlite3_prepare_*`](https://sqlite.org/c3ref/prepare.html)
2. Bind input via [`sqlite3_bind_*`](https://sqlite.org/c3ref/bind_blob.html)
3. Call [`sqlite3_step`](https://sqlite.org/c3ref/step.html) to start processing the statement
4. Call for each column [`sqlite3_column_value`](https://sqlite.org/c3ref/column_blob.html) to get a pointer to the corresponding column inside of the sqlite statement.
5. Repeat 3. + 4. for all rows 

The problem here is that step 5. will invalid the pointers received in step 4. They now will point to the corresponding values for the new row. But if we use the `Iterator` interface here users could hold a previous row and try to access those values. Now we could say we just cannot use `Iterator` here as that's done by the other database libraries. Or we could do some clever stuff. Basically we return an `Rc<…>` as row. That allows us to hold an reference to the previous row inside of our iterator implementation. We then can check if someone holds the previous row or not by just checking the strong count of the `Rc<…>`. If that count is 1 our internal instance is the only reference to the previous row. That means  we can just overwrite the content there and reuse the allocation. That's the happy path.
If the strong count is not 1, someone else has a reference to the previous row. Now as the row is internally `Rc<RefCell<_>>` we are able to modify the content of their row behind the scenes. We do that by just cloning the corresponding column values for this last row and injecting the cloned values in the last row. In that way we can forward the statement iterator another step as the value ptrs are now not used by the previous row. That's the duplicated_row_case. I assume that this is the unlikely path, therefore the separate function + `#[cold]`. This assumption is based on the observation that diesels own `RunQueryDsl` basically always hits the happy path, as we don't hold an reference to the last row at all. To hit the other path users must directly call `Connection::load()` and store the rows somewhere.

Reply:
Thank you so much for your answer! I learned a little bit more about unsafe code and use cases for `Rc` and `RefCell` structs. I'm still a bit confused overall especially on this point

> That allows us to hold an reference to the previous row inside of our iterator implementation. We then can check if someone holds the previous row or not by just checking the strong count of the Rc<…>.

Like... who else would hold this value? If `diesel` is the only thing responsible for iterating through the returned data from the sqlite driver, then why would this be needed? EDIT: Now actually I see "To hit the other path users must directly call Connection::load() and store the rows somewhere." I think I may have answered my own question 😅 


Also I think I don't understand statements very well; my understanding was always that a prepared statement was something the client sent to the database to say "hey expect that the signature of this query I'm about to want to run is X" and then when we actually send along the parameters we're basically delegating to the database to do sanitization and data type assertions (to avoid SQL injection); so I'm a bit confused when it seems like we actually use [the statement struct itself to store the values (by values I mean the actual integers, strings, etc. that are returned for each row) themselves](https://github.com/diesel-rs/diesel/blob/c237a74b63900b734c659a201fc1d6b87108033b/diesel/src/sqlite/connection/statement_iterator.rs#L49-L60)

Reply:
Oh and by the way, now that I've sat on this some more I think I see where the complexity is; and that's in the value of each row needing to be stored via some pointer that is perpetually written and re-written -- I guess that's very memory efficient but also potentially disastrous in case multiple tasks are trying to read the value at that point in a concurrent manner and have race conditions (which it seems like `diesel` is handling intelligently via the reference counting... although I wonder now why `Rd` and not `Arc` but that's a separate issue -- I've learned enough today :D)

Reply:
> why `Rc` and not `Arc`

`Rc<RefCell>` would have to be replaced by `Arc<Mutex>` to allow the user to send it across threads, the performance overhead of which seems not worth considering how unlikely it is that someone actually needs this feature:

> diesels own `RunQueryDsl` basically always hits the happy path, as we don't hold an reference to the last row at all

Reply:
Thanks! Still trying to pick up pieces of diesel and I'll look into the `RunQueryDsl` trait next :)

Reply:
> Also I think I don't understand statements very well; my understanding was always that a prepared statement was something the client sent to the database to say "hey expect that the signature of this query I'm about to want to run is X" and then when we actually send along the parameters we're basically delegating to the database to do sanitization and data type assertions (to avoid SQL injection); so I'm a bit confused when it seems like we actually use the statement struct itself to store the values (by values I mean the actual integers, strings, etc. that are returned for each row) themselves

A prepared statement does more than just checking and serializing bind value. It's a way to tell the server that we want to execute a certain query with the following variables. What a clever optimizing server now could do is to generate an executable optimized representation of that query internally and link it to the prepared statement. In the end SQL is just an interpreted programming language, so the server may prepare that early on. As a client library we can use this behaviour in various ways:

1. We can delegate bind serialization + sanitizing to the server, as prepared statements allow us to use place holder values
2. We can store references to the prepared statements that are likely to be reexecuted later on. This saves the amount of time needed to prepare the statement on query reexecution. (Checkout the [`diesel::connection::statement_cache`](https://docs.diesel.rs/master/diesel/connection/statement_cache/index.html) module in diesel for details about how this is handled in diesel)

Depending on the actual database implementation prepared statements could bundle more functionality. For SQLite a statement could be executed and the execution is tracked inside the statement itself. That seems to be meaningful if you think about a prepared statement as a function/state combination that represents some SQL query. A SQLite prepared statements is basically a program that waits to be executed. You can start the execution at any time and for a stared execution it will track the current state. If required you can reset it to the starting point by using [`sqlite3_reset`](https://sqlite.org/c3ref/reset.html)

> > why Rc and not Arc
>
> Rc<RefCell> would have to be replaced by Arc<Mutex> to allow the user to send it across threads, the performance overhead of which seems not worth considering how unlikely it is that someone actually needs this feature:

The main reason to not using a `Arc<Mutex<_>>` there is that it is really hard to use `SqliteConnection` in multiple threads at the same time at all. The type implements only `Send`, but not `Sync`. That means you cannot execute queries from different threads using the same connection at the same time. This also applies to the returned iterator. You cannot call next from different threads at the same time anyway. That's the main reason to use `Rc<RefCell<_>>` instead of `Arc<Mutex<_>>`.